src/dmd/dtemplate.d

```diff
@@ -8039,7 +8043,16 @@ MATCH matchArg(TemplateParameter tp, Scope* sc, RootObject oarg, size_t i, Templ
         }
 
         //printf("\tvalType: %s, ty = %d\n", tvp.valType.toChars(), tvp.valType.ty);
-        vt = tvp.valType.typeSemantic(tvp.loc, sc);
+
+        // The type might rely on preceeding parameters, e.g.
+        // void foo(T, T val)(...) {}
+        // Make a copy to not modify the template declaration!
```

Comment:
Reference to bugzilla would be good + a note about the issue itself i.e. the emphasis is sort of there with the "!" but this smells of something really broken in the sema code somewhere so best make it super obvious.

Reply:
> but this smells of something really broken in the sema code somewhere so best make it super obvious.

Not necessarily, the assumption for template instances is to work on an independent copy of the AST that can be modified by semantic. The code previously didn't make a copy at all.

Making copy conditional is mostly intended to avoid redundant copies.

Reply:
Looks like my email reply got mangled oof, anyway what I meant to say was that it should still be noted in the source code. if there is one thing dmd does not have too much of its prose.